### PR TITLE
E2E Integration Testing Framework

### LINKED ISSUE #73: No E2E Integration Between Pessimism Tests & EVM Execution Nodes
## Problem
There currently exists no end-to-end integration tests that validate Pessimism's correctness. Currently, the only way to validate the e2e (ie. `block --> invariantInput-->activation-->alert`) flow is via manual tests on a local instance of the application. This is problematic as:
I. Code changes that can impact correctness aren't programmatically assessed when being considered for merge
II. Performing manual validations is exhaustive and disparate from the tests in the actual repo

## Problem Solution
* **[1]** (https://github.com/base-org/pessimism/pull/74) Re-architect the existing main caller abstractions into an `Application` struct that has accessor methods for starting, stopping, and calling application level methods. The `Application` should hold all three subsystems the http server.
* **[2]** (https://github.com/base-org/pessimism/pull/85) Develop a way to register downstream alert invocations. A huge part of testing will require validating that an alert actually fires given an expected sequence of blocks and transactions. 
* **[3]** (https://github.com/base-org/pessimism/pull/85) Develop a testing framework that assess that correctness of the `balance_enforcement` and `contract_event` invariants. The existing Optimism [monorepo](https://github.com/ethereum-optimism/optimism/tree/develop) already has a op-geth node [representation](https://github.com/ethereum-optimism/optimism/blob/develop/op-e2e/op_geth.go) that can be leveraged for these tests.

### PR DESCRIPTION
**Pull request by epociask**
## Fixes Issue

Fixes #73 
## Changes proposed

* Reworked application context to store `L1Client` and `L2Client` interfaces. This ensure that only one instance of each client is used throughout the application. Initially, we would create a new one in the component constructor. 
* Added an `e2e` package that stores all integration tests
  - Balance enforcement invariant test
  - Http test server to represent a slack server
  - Use of `op-e2e` to leverage existing bedrock testing framework
* Added `e2e-test` github workflow to run integration tests
* Updated `/health` service tests 

## Note to reviewers
* Most the changes in this PR are in `go.mod` & `go.sum` dependencies

### CODE REVIEW

File: internal/api/service/health.go [hunk 1]
```diff
@@ -4,41 +4,43 @@ import (
...
 // CheckHealth ... Returns health check for server
 func (svc *PessimismService) CheckHealth() *models.HealthCheck {
 	hc := &models.ChainConnectionStatus{
-		IsL1Healthy: svc.CheckETHRPCHealth(svc.cfg.L1RpcEndpoint),
-		IsL2Healthy: svc.CheckETHRPCHealth(svc.cfg.L2RpcEndpoint),
+		IsL1Healthy: svc.CheckETHRPCHealth(core.Layer1),
+		IsL2Healthy: svc.CheckETHRPCHealth(core.Layer2),
 	}
 
 	healthy := hc.IsL1Healthy && hc.IsL2Healthy
```
Comment (anupsv-cb) [inline internal/api/service/health.go:21]: A thought, is it worth parallelizing the health check here ? Would we be using this endpoint for aws health check ?

File: cmd/pessimism/main.go [hunk 2]
```diff
@@ -4,133 +4,67 @@ import (
...
+// main ... Application driver
+func main() {
+	cfg := config.NewConfig(cfgPath) // Load env vars
+	ctx := context.Background()      // Create context
+	logger := logging.WithContext(ctx)
+
+	l1Client, err := client.NewEthClient(ctx, cfg.L1RpcEndpoint)
 	if err != nil {
-		return nil, nil, err
+		logging.WithContext(ctx).Fatal("Error creating L1 client", zap.Error(err))
```
Comment (adrain-cb) [inline cmd/pessimism/main.go:29]:
```suggestion
		logger.Fatal("Error creating L1 client", zap.Error(err))
```

File: cmd/pessimism/main.go [hunk 3]
```diff
@@ -4,133 +4,67 @@ import (
...
 	}
 
-	server, cleanup, err := server.New(ctx, cfg.ServerConfig, handler)
+	l2Client, err := client.NewEthClient(ctx, cfg.L2RpcEndpoint)
 	if err != nil {
-		return nil, nil, err
+		logging.WithContext(ctx).Fatal("Error creating L1 client", zap.Error(err))
```
Comment (adrain-cb) [inline cmd/pessimism/main.go:34]:
```suggestion
	logger.Fatal("Error creating L1 client", zap.Error(err))
```

File: internal/api/service/health.go (hunk 1 above)
Comment (adrain-cb) [reply internal/api/service/health.go:21]: +1; would definitely make the healthcheck respond faster

Comment (epociask) [reply internal/api/service/health.go:22]: #88

File: cmd/pessimism/main.go (hunk 3 above)
Comment (epociask) [reply cmd/pessimism/main.go]: [2ab1601](https://github.com/base-org/pessimism/pull/85/commits/2ab16013e3d520d234c6f3a0c27d9a51ef596a07)

File: cmd/pessimism/main.go (hunk 2 above)
Comment (epociask) [reply cmd/pessimism/main.go]: [2ab1601](https://github.com/base-org/pessimism/pull/85/commits/2ab16013e3d520d234c6f3a0c27d9a51ef596a07)

**Review by adrain-cb — DISMISSED** [review]

**Review by nadir-akhtar-coinbase — APPROVED** [review]

**Review by nadir-akhtar-coinbase — DISMISSED** [review]

**Review by nadir-akhtar-coinbase — APPROVED** [review]

**Pull request merged**